[RFC] scripts(build/{termux_step_override_config_scripts,toolchain/23c,toolchain/27b}): fix crossbuild prefix pollution

robertkirkman opened the pull request:
Fixes #20336 and progress on #21130 

Fixes

- `ERROR: Failed running '/data/data/com.termux/files/usr/bin/llvm-config', binary or interpreter not executable.`
- `Interpreter: Cannot run the interpreter "/data/data/com.termux/files/usr/bin/python3"` 

in `mesa`  + a large number of other packages that I forgot, when built from entrypoint `scripts/run-docker.sh ./build-all.sh`

Fixes many errors similar to

- `nasmlib/md5c.c:46:10: error: no member named 'buf' in 'struct MD5Context'`

in

- `bitlbee`
- `emacs`
- `gdb`
- `libjxl`
- `libmediainfo`
- `maxcso`
- `nodejs`
- `pipewire`
- `rust`
- `sqlcipher`
- `tome2`
- `weggli`
- `zip`

when built from entrypoint `scripts/run-docker.sh ./build-all.sh`.

I have **not finished testing** all the codepaths I want to test these changes on,
but I would like to know, do you know any better solutions to these errors or better ways to write these changes?

[The commit guidelines say you have an 80 column limit](https://github.com/termux/termux-packages/blob/497c74e23d0fa8c93d990b69108228affc1cd527/CONTRIBUTING.md#commit-guidelines), but the same file I edited already has multiple 120+ column lines, leaving me confused. Should I reformat the entire file to fit in 80 columns?

Also, the commit guidelines do not have very many directions regarding changes to the `scripts` folder, leaving me unsure whether I'm following the correct coding style in general for this folder. 

### Possibly preexisting alternative solution

I strongly suspect that [termux-play-store/termux-packages](https://github.com/termux-play-store/termux-packages/commit/2bef6d4591e1ab5e0ba3e588a5ce19559bdd3d1e) works around/bypasses many of these errors using a very different codepath. It is possible that maybe fornwall prefers that method over this one, so if that repo is going to combine into this repo someday, maybe it's better to use that solution instead.

fornwall commented:
Regarding https://github.com/termux-play-store/termux-packages/commit/2bef6d4591e1ab5e0ba3e588a5ce19559bdd3d1e - always starting `$PREFIX` from a clean slate and install dependencies from .deb/.pkg packages (regardless if they are built locally or not) - I think that's a good approach in general, to simplify reasoning about builds and making them more reproducible, instead of the build result depending on what might be left behind from previous (related or unrelated) builds. It currently takes some shortcuts and needs to be cleaned up and generalized to pacman and more build options, but I intend to submit a PR for discussion soon.

But regardless, this PR is interesting! It's not actually clear to me why we see a lot of build errors such as the ones above:

- `ERROR: Failed running '/data/data/com.termux/files/usr/bin/llvm-config', binary or interpreter not executable.`
- `Interpreter: Cannot run the interpreter "/data/data/com.termux/files/usr/bin/python3"`

when building with `./build-all.sh`, but we do not get the error when building the package directly with `./build-package.sh -i <pkg>`. Is that because (in the above two examples of errors) the dependencies does not include `$PREFIX/bin/llvm-config` or `$PREFIX/bin/python3`, so they are just not there interferring with the build?

And in packages that does include cross-compiled `$PREFIX/bin/llvm-config` or `$PREFIX/bin/python3` installed through package depdencies, so that the files are there when building with `-i`, the build just happens to work (or we have worked around it in the build, like specifying `-DPYTHON_EXECUTABLE=$(command -v python3)` to cmake, to pick up the host `python3` instead of the cross-compiled one)?

Put in another way: If we migrate `build-all.sh` to always start out with a clean `$PREFIX` and install dependencies from locally built packages when building each package, are the changes in this PR necessary and/or desireable?

Maxython commented on scripts/build/termux_step_override_config_scripts.sh:
```diff
@@ -3,11 +3,28 @@ termux_step_override_config_scripts() {
 	ln -sf /bin/sh "$TERMUX_PREFIX/bin/sh"
 
-	if [ "$TERMUX_INSTALL_DEPS" = false ] || [ "$TERMUX_PACKAGE_LIBRARY" = "glibc" ]; then
+	if [ "$TERMUX_PACKAGE_LIBRARY" = "glibc" ]; then
```
Please move this logic to the `[ "$TERMUX_ON_DEVICE_BUILD" = true ]` logic which is at the beginning of the function.
```bash
if [ "$TERMUX_ON_DEVICE_BUILD" = true ] || [ "$TERMUX_PACKAGE_LIBRARY" = "glibc" ]; then
	return
fi
```

robertkirkman replied:
Oh ok, so to clarify that as much as possible, you are saying that the prexisting line `ln -sf /bin/sh "$TERMUX_PREFIX/bin/sh"` should _not_ be executed if `TERMUX_PACKAGE_LIBRARY` is `glibc`? 

That would be a removing preexisting code from the glibc codepath, which I have not tested. Would you like me to test building all glibc packages with this PR as well?

Maxython replied:
> Oh ok, so to clarify that as much as possible, you are saying that the preexisting line ln -sf /bin/sh "$TERMUX_PREFIX/bin/sh" should not be executed if TERMUX_PACKAGE_LIBRARY is glibc?

Yes and no. Your change now replaces not only `sh`, but all possible binaries/scripts from Termux packages. Because of this, I have a concern that if the build script does not skip this stage when compiling for glibc packages, it may break the build of some packages. To understand: in some glibc packages, binaries or scripts from glibc packages may be used (i.e. runned) for compilation.

robertkirkman replied:
> all possible binaries/scripts from Termux packages

This part ensures that _only_ files that are _not_ identified as "text" by `file` command are modified

but I understand, if the glibc packages need to use the `aarch64` binaries directly during their builds, it needs to be like you suggested. I applied that change in my recent push.

robertkirkman commented:
> Is that because (in the above two examples of errors) the dependencies does not include $PREFIX/bin/llvm-config or $PREFIX/bin/python3, so they are just not there interferring with the build when building with -i?

Correct, for the Python3 example and some other similar errors this fixes, but in the case of specifically `llvm-config`, it is a subtly different problem that was reported in, and which I explained in some additional detail in, #20336 . Basically, there is a condition `if [ "$TERMUX_INSTALL_DEPS" = false ]` that was causing an error for me and other people, including it would seem, the people in #21130 . Do you know what the original purpose of that condition is and what package it's needed for? For me it seems to work better when I remove it.

> Put in another way: If we migrate build-all.sh to always start out with a clean $PREFIX and install dependencies from locally built packages when building each package, are the changes in this PR necessary and/or desireable?

No, if we wait for your PR and go with that, I'm pretty sure it avoids all errors caused by the prefix having other packages installed into it previously (at a very small, surely insignificant speed cost incurred by repeatedly reconstructing the prefix every time a package builds)

I use Gentoo, so every time I update my PC, every new package installed is recompiled directly using the same rootfs that I have kept installed since I installed the OS. I guess my method in this PR is just inspired by that, to be able to continuously keep rebuilding and reinstalling all Termux packages in an endless loop, into the same crossbuild prefix, by attempting to universally solve all potential errors that can occur when building the repository that way.

robertkirkman commented:
As a preview and additional example of the types of things that might need to be touched if I continue down the pattern this PR follows, I'm pretty sure that when the prefix gets into a certain state (I forgot exactly how to get to that state but my docker container will probably cycle back around to it eventually), this, and a few similar changes, were required to avoid more errors, due to the eventual spontaneous appearance of a `libz.a`:

```diff
--- a/packages/cargo-c/build.sh
+++ b/packages/cargo-c/build.sh
@@ -41,11 +41,14 @@ termux_step_pre_configure() {
 
 	mv $TERMUX_PREFIX/lib/libz.so.1{,.tmp}
 	mv $TERMUX_PREFIX/lib/libz.so{,.tmp}
+	mv $TERMUX_PREFIX/lib/libz.a{,.tmp}
 
 	ln -sfT $(readlink -f $TERMUX_PREFIX/lib/libz.so.1.tmp) \
 		$_CARGO_TARGET_LIBDIR/libz.so.1
 	ln -sfT $(readlink -f $TERMUX_PREFIX/lib/libz.so.tmp) \
 		$_CARGO_TARGET_LIBDIR/libz.so
+	ln -sfT $(readlink -f $TERMUX_PREFIX/lib/libz.a.tmp) \
+		$_CARGO_TARGET_LIBDIR/libz.a
 
 	if [[ "${TERMUX_ARCH}" == "x86_64" ]]; then
 		RUSTFLAGS+=" -C link-arg=$($CC -print-libgcc-file-name)"
@@ -55,9 +58,11 @@ termux_step_pre_configure() {
 termux_step_post_make_install() {
 	mv $TERMUX_PREFIX/lib/libz.so.1{.tmp,}
 	mv $TERMUX_PREFIX/lib/libz.so{.tmp,}
+	mv $TERMUX_PREFIX/lib/libz.a{.tmp,}
 }
 
 termux_step_post_massage() {
 	rm -f lib/libz.so.1
 	rm -f lib/libz.so
+	rm -f lib/libz.a
 }
```
Before I copy and paste this part ^ into this PR, I'm going to check the error it was needed to solve again, and think about rewriting it or coming up with an even better solution for when there are Library file conflicts.

So, a TL;DR summary of what's going on here is,

- fixing the `$TERMUX_PREFIX/bin` folder for endless cross-compiling of all packages
- fixing the `$TERMUX_PREFIX/include` folder for endless cross-compiling of all packages
- possibly TODO: fix the `$TERMUX_PREFIX/lib` folder for endless cross-compiling of all packages?

TomJo2000 commented:
I just remembered we have an open PR for updating the NDK 23c patches.
I have a feeling that might have implications for this PR or vise versa.

I just hadn't had the time to review the earlier PR so far.
- #21499

robertkirkman commented:
> I just remembered we have an open PR for updating the NDK 23c patches. I have a feeling that might have implications for this PR or vise versa.
> 
> I just hadn't had the time to review the earlier PR so far.
> 
>     * [Update NDK 23c patches #21499](https://github.com/termux/termux-packages/pull/21499)

The purpose of changing `-I` to `-isystem` in those files is because , according to the [clang documentation](https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-I-dir), "if there are multiple -I options, these directories are searched in the order they are given before the standard system directories are searched. If the same directory is in the SYSTEM include search paths, for example if also specified with -isystem, the -I option will be ignored" 

That means that:

- before my change, `/data/data/com.termux/files/usr/include` was searched by the cross-compiler for headers _before_ attempting to detect headers that were _also_ specified with `-I` within the build system internal to each package being compiled. That was the cause of `nasmlib/md5c.c:46:10: error: no member named 'buf' in 'struct MD5Context'` (because the `md5.h` and other headers from `libmd` package [and other affected packages] were being wrongly included into the build of all packages that contain their own internal `md5.h`).
- after my change, `/data/data/com.termux/files/usr/include` is searched by the cross-compiler for headers _after_ **all** instances of `-I` passed by the build system internal to the package are searched. I believe this consistently results in more reliable header behavior.

twaik commented:
> > `nasmlib/md5c.c:46:10: error: no member named 'buf' in 'struct MD5Context'`
> 
> I think I recall seeing a very similar error with `handbrake` as well. Couldn't pin it down there, good to know that this is the cause.

It happens because `libmd` installs `md5.h` to `$TERMUX_PREFIX/include` and there is `-I$TERMUX_PREFIX/include` in `$CPPFLAGS` so `$TERMUX_PREFIX/include` has higher precedence than include directories in build folders.

robertkirkman commented:
@twaik I decided to continue discussion and documenting some more test results regarding the things like 

- `-sysroot` (I believe it affects `lib` + `include` folders), 
- `-isysroot`/`-isystem` (I believe these  could affect the `include` folder only) 
- or `-DDEFAULT_SYSROOT` (affects a resulting `clang` program using the compile-time settings of the compiler being built) 

here since this might be a more appropriate thread to talk about it, since I mentioned something about the sysroot already above. The `clang` program that comes inside Termux itself was compiled with `-DDEFAULT_SYSROOT`,  and the cross-compiler that comes from the official NDK was not, so one or two of the subtle differences in behavior between the cross-compiler and the Termux app can be pinpointed to this detail.

This additional explanation might help reveal **where did some of my ideas come from when planning changes in this PR**. I did not completely invent these ideas by myself, intstead, I happened to be observing the behavior of several toolchains and noticing what they do in various situations, and I have sometimes been taking inspiration from or copying what they do in order to fix some errors that happened when building packages in termux-packages.

### Gentoo amd64

Gentoo has a package called `crossdev` that allowed me to easily install a regular GNU/Linux cross-compiler for other projects that do not involve Android. It was compiled using a GCC equivalent to Clang's `-DDEFAULT_SYSROOT`, which sets the prefix to the correct folder for cross-compiling Gentoo packages.
```bash
tacokoneko@CORSAIR ~ $ aarch64-none-linux-gnu-gcc -print-sysroot
/usr/aarch64-none-linux-gnu
tacokoneko@CORSAIR ~ $ ls /usr/aarch64-none-linux-gnu
etc  lib  lib64  sbin  sys-include  usr  var
```


### Termux aarch64
`--sysroot=/data/data/com.termux/files` is present in the output of a `clang` debugging info command.

```bash
tacokoneko@CORSAIR ~ $ ssh -p 8022 192.168.12.191
tacokoneko@192.168.12.191's password: 
Welcome to Termux!

Docs:       https://termux.dev/docs
Donate:     https://termux.dev/donate
Community:  https://termux.dev/community

Working with packages:

 - Search:  pkg search <query>
 - Install: pkg install <package>
 - Upgrade: pkg upgrade

Subscribing to additional repositories:

 - Root:    pkg install root-repo
 - X11:     pkg install x11-repo

For fixing any repository issues,
try 'termux-change-repo' command.

Report issues at https://termux.dev/issues
~ $ echo '' | clang -x c - -v 2>&1 | grep -e "--sysroot="
 "/data/data/com.termux/files/usr/bin/ld.lld" --sysroot=/data/data/com.termux/files -EL --fix-cortex-a53-843419 -z now -z relro -z max-page-size=16384 --hash-style=gnu -rpath=/data/data/com.termux/files/usr/lib --eh-frame-hdr -m aarch64linux -pie -dynamic-linker /system/bin/linker64 -o a.out /data/data/com.termux/files/usr/lib/crtbegin_dynamic.o -L/data/data/com.termux/files/usr/lib -L/data/data/com.termux/files/usr/aarch64-linux-android/lib -L/system/lib64 /data/data/com.termux/files/usr/tmp/--88abe0.o /data/data/com.termux/files/usr/lib/clang/19/lib/linux/libclang_rt.builtins-aarch64-android.a -l:libunwind.a -ldl -lc /data/data/com.termux/files/usr/lib/clang/19/lib/linux/libclang_rt.builtins-aarch64-android.a -l:libunwind.a -ldl /data/data/com.termux/files/usr/lib/crtend_android.o
~ $ 
logout
Connection to 192.168.12.191 closed.
```

### Termux package builder Docker image

On the other hand, that argument does not show up in this compiler because it is a prebuilt compiler that comes from a non-Termux source. This means that, technically, to precisely synchronize the exact literal behavior of the cross-compiler and the non-cross-compiler, either `-DDEFAULT_SYSROOT` must be removed from the build of the `clang` package, or `-DDEFAULT_SYSROOT` must be added to the build of a custom built NDK specifically for cross-compiling Termux packages. That is just an example of a very overly invasive solution though, I think it is probably unnecessary to recompile the entire cross-compiler, and a similar result can probably be achieved by setting up the `--sysroot` argument or similar arguments passed to the current build of the cross-compiler.
```bash
tacokoneko@CORSAIR ~ $ code/termux/electric-boogaloo/termux-packages/scripts/run-docker.sh 
Running container 'termux-package-builder' from image 'temporary-local-termux-package-builder-image'...
builder@bd15ad372c73:~/termux-packages$ echo '' | /home/builder/.termux-build/_cache/android-r27b-api-24-v1/bin//aarch64-linux-android-clang -x c - -v 2>&1 | grep -e "--sysroot="
builder@bd15ad372c73:~/termux-packages$ 
```

twaik commented:
AFAIK NDK's clang puts --sysroot argument automatically when you pass `--target` argument.

robertkirkman commented:
Good idea, though in my command shown, the result is the same as the equivalent with `--target` because of the script in that folder that i invoked already containing `--target`. however it might be a good idea for me to save the full result here that I see when I use this command within the docker container:

```
echo '' |  /home/builder/.termux-build/_cache/android-r27b-api-24-v1/bin/clang --target=aarch64-linux-android24 -x c - -v
```

```
Android (12285214, +pgo, +bolt, +lto, +mlgo, based on r522817b) clang version 18.0.2 (https://android.googlesource.com/toolchain/llvm-project d8003a456d14a3deb8054cdaa529ffbf02d9b262)
Target: aarch64-unknown-linux-android24
Thread model: posix
InstalledDir: /home/builder/.termux-build/_cache/android-r27b-api-24-v1/bin
 "/home/builder/.termux-build/_cache/android-r27b-api-24-v1/bin/clang-18" -cc1 -triple aarch64-unknown-linux-android24 -emit-obj -mrelax-all -dumpdir a- -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name - -mrelocation-model pic -pic-level 2 -pic-is-pie -mframe-pointer=non-leaf -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu generic -target-feature +neon -target-feature +v8a -target-feature +fix-cortex-a53-835769 -target-abi aapcs -debugger-tuning=gdb -fdebug-compilation-dir=/home/builder/termux-packages -v -fcoverage-compilation-dir=/home/builder/termux-packages -resource-dir /home/builder/.termux-build/_cache/android-r27b-api-24-v1/lib/clang/18 -internal-isystem /home/builder/.termux-build/_cache/android-r27b-api-24-v1/lib/clang/18/include -internal-isystem /home/builder/.termux-build/_cache/android-r27b-api-24-v1/bin/../sysroot/usr/local/include -internal-externc-isystem /home/builder/.termux-build/_cache/android-r27b-api-24-v1/bin/../sysroot/usr/include/aarch64-linux-android -internal-externc-isystem /home/builder/.termux-build/_cache/android-r27b-api-24-v1/bin/../sysroot/include -internal-externc-isystem /home/builder/.termux-build/_cache/android-r27b-api-24-v1/bin/../sysroot/usr/include -ferror-limit 19 -femulated-tls -fno-signed-char -fgnuc-version=4.2.1 -fcolor-diagnostics -target-feature +outline-atomics -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /tmp/--eede17.o -x c -
clang -cc1 version 18.0.2 based upon LLVM 18.0.2 default target x86_64-unknown-linux-gnu
ignoring nonexistent directory "/home/builder/.termux-build/_cache/android-r27b-api-24-v1/bin/../sysroot/usr/local/include"
ignoring nonexistent directory "/home/builder/.termux-build/_cache/android-r27b-api-24-v1/bin/../sysroot/include"
#include "..." search starts here:
#include <...> search starts here:
 /home/builder/.termux-build/_cache/android-r27b-api-24-v1/lib/clang/18/include
 /home/builder/.termux-build/_cache/android-r27b-api-24-v1/bin/../sysroot/usr/include/aarch64-linux-android
 /home/builder/.termux-build/_cache/android-r27b-api-24-v1/bin/../sysroot/usr/include
End of search list.
 "/home/builder/.termux-build/_cache/android-r27b-api-24-v1/bin/ld.lld" -EL --fix-cortex-a53-843419 -z now -z relro -z max-page-size=4096 --hash-style=gnu --eh-frame-hdr -m aarch64linux -pie -dynamic-linker /system/bin/linker64 -o a.out /home/builder/.termux-build/_cache/android-r27b-api-24-v1/bin/../sysroot/usr/lib/aarch64-linux-android/24/crtbegin_dynamic.o -L/home/builder/.termux-build/_cache/android-r27b-api-24-v1/lib/clang/18/lib/linux/aarch64 -L/home/builder/.termux-build/_cache/android-r27b-api-24-v1/bin/../sysroot/usr/lib/aarch64-linux-android/24 -L/home/builder/.termux-build/_cache/android-r27b-api-24-v1/bin/../sysroot/usr/lib/aarch64-linux-android -L/home/builder/.termux-build/_cache/android-r27b-api-24-v1/bin/../sysroot/usr/lib /tmp/--eede17.o /home/builder/.termux-build/_cache/android-r27b-api-24-v1/lib/clang/18/lib/linux/libclang_rt.builtins-aarch64-android.a -l:libunwind.a -ldl -lc /home/builder/.termux-build/_cache/android-r27b-api-24-v1/lib/clang/18/lib/linux/libclang_rt.builtins-aarch64-android.a -l:libunwind.a -ldl /home/builder/.termux-build/_cache/android-r27b-api-24-v1/bin/../sysroot/usr/lib/aarch64-linux-android/24/crtend_android.o
ld.lld: error: undefined symbol: main
>>> referenced by crtbegin.c
>>>               /home/builder/.termux-build/_cache/android-r27b-api-24-v1/bin/../sysroot/usr/lib/aarch64-linux-android/24/crtbegin_dynamic.o:(_start_main)
>>> referenced by crtbegin.c
>>>               /home/builder/.termux-build/_cache/android-r27b-api-24-v1/bin/../sysroot/usr/lib/aarch64-linux-android/24/crtbegin_dynamic.o:(_start_main)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

It seems to have set in it a large number of arguments to use a relative path to set the sysroot at the folder above its folder, i.e. ` /home/builder/.termux-build/_cache/android-r27b-api-24-v1/`, but without any of them being exactly `--sysroot`.  

It is not completely clear to me whether using `-isystem`, for example, would risk overwriting the path this compiler sets for -internal-externc-isystem`  and causing an error if the headers in that path are needed, or not, but I can continue testing and check whether there seems to be any risk of that.

twaik commented:
Probably `-internal-externc-isystem` has the lowest priority and considered to be invoked only by clang itself. Idk.

robertkirkman commented:
Sorry if it is too much flooding to continue posting notes like this, but I would like to mention since I just reached the `proxmark3` package right now, that, while my current `-isystem` argument fixes a lot of packages without further intervention to modify the package, there are some unique packages that, at least for now, require their own toolchain setup variables, and the way some of them are prevents my `$CPPFLAGS` from propagating into their builds, meaning that my solution cannot penetrate into their builds to fix their manifestations of the `libmd`/other headers errors without changing them a little bit as well.

https://github.com/termux/termux-packages/blob/ac069d232bbb71b95ce0f9405ed7abd4eb5b8aa5/packages/proxmark3/build.sh#L13-L17

```
# this is how to shortcut reproduce libmd-related errors like the one you are familiar with, without having
# to run the entire build-all.sh from scratch every time. note that it is not always libmd sometimes it is other
# packages, so for similar errors, it's necessary to correctly identify the package that comes first in the pollution order
# before it can be reproduced this way. In this specific case it is libmd.
scripts/run-docker.sh ./build-package.sh -I libmd proxmark3
```

- `src/cmdflashmem.c:81:5: error: call to undeclared function 'mbedtls_sha1'`

However, what seems like an acceptable solution in this particular case is to adjust the package like this,

```diff
--- a/packages/proxmark3/build.sh
+++ b/packages/proxmark3/build.sh
@@ -11,8 +11,8 @@ TERMUX_PKG_BUILD_IN_SRC="true"
 TERMUX_PKG_BLACKLISTED_ARCHES="i686, x86_64"
 
 termux_step_post_configure() {
-       export LDLIBS="-L${TERMUX_PREFIX}/lib"
-       export INCLUDES="-I${TERMUX_PREFIX}/include"
+       export LDLIBS="$LDFLAGS"
+       export INCLUDES="$CPPFLAGS"
        TERMUX_PKG_EXTRA_MAKE_ARGS="client CC=$CC CXX=$CXX LD=$CXX cpu_arch=$TERMUX_ARCH SKIPREVENGTEST=1 SKIPQT=1 SKIPPTHREAD=1 SKIPGD=1 PLATFORM=PM3GENERIC"
 }
 ```

That allows the command `scripts/run-docker.sh ./build-package.sh -I libmd proxmark3` to complete successfully for me, when combined with my other change in `termux_setup_toolchain_27b.sh` (it also allows me to resume and continue the same run of `build-all.sh` I was running where the error first appeared for me)

I'll probably commit that to this PR soon as my change for `proxmark3`.

twaik commented:
> ```shell
> # note that it is not always libmd sometimes it is other
> # packages, so for similar errors, it's necessary to correctly identify the 
> # package that comes first in the pollution order
> # before it can be reproduced this way. In this specific case it is libmd.
> ```

It is easy. Make and ninja print the command which fail. You simply add termux's toolchain path to PATH and navigate to build folder like 
```
export PATH="~/.termux-build/_cache/android-r27b-api-24-v1/bin:$PATH"
cd ~/.termux-build/nasm/src # because it builds in SRCDIR...
```
and invoke the failing clang command with `-H` in CFLAGS and it will print all the headers it includes with their real paths.

After this you only should the package providing the file with `apt-file search /data/data/com.termux/files/usr/include/md5.h` (in termux environment).

robertkirkman commented:
That was really weird, I pushed a change to the `dropbear` package that seemed very small and fixed the build for me locally, but then in CI it failed, but then when I force pushed the exact same change again it did not fail. The error in CI was really weird, it said `/home/runner/work/termux-packages/termux-packages/packages/composer/build.sh: line 14: composer: command not found` even though I did not do anything with the `composer` package. If my change to `dropbear` causes an intermittent failure in CI that would not be good, so I left a note about it.

This was the run that failed and then when I deleted the commit, re-committed and force pushed without making any code changes it succeeded. https://github.com/termux/termux-packages/actions/runs/11582453980

twaik commented:
Probably it is time for final review for this PR. We have some issue with updating/uploading a lot of packages at once.

robertkirkman commented:
> Probably it is time for final review for this PR. We have some issue with updating/uploading a lot of packages at once.

Ok, in that case I need to think about how to organize it into several separate PRs I think, that are smaller and more pinpointed into the exact changes they are relevant to. For example, the whole large blue Note at the beginning could be separated out into its own PR, and the code associated with it is ready. 

On the other hand, several of the other changes besides that one are not ready yet, so rather than continuing to lump all of my changes of the type that you usually describe as "`build-all.sh` changes" into this one, I need to separate them into their own specific categories.

One of the reasons I decide to mark my own `build-all.sh` changes as a draft is because, **in my estimations of likely outcomes, I believe the only way to exhaustively detect and solve all present and even some future errors is by running the entire `build-all.sh` twice over the same `$TERMUX_PREFIX`.** That means that I would consider some of these changes fully tested once my container has already successfully compiled all packages at least once, then reset the build status without resetting the whole prefix and also successfully compiled all packages _again_ a second time.

The reason I believe that is because, it is my understanding that gradual replacement of dependencies in packages over time, as they receive updates, can subtly shift the **build order calculated by `build-all.sh`.** That could hypothetically lead to a situation in the future when a package that `build-all.sh` did not previously compile before another package on the 1st run, could shift in the build order to before, exposing untested edge cases and potentially leading to more errors. I believe that if I run `build-all.sh` 2 entire times, it will allow me to find preemptively prevent all of those potential future errors.

I definitely want to try to prevent as many potential errors in my code as possible by fully testing all codepaths.

twaik commented:
On the one hand you are right, it is good to commit all changes at once.
On the other hand we can get into situation when we have multiple fixes of the same problems from different developers, and all these developers spend significant amount of time to create fixes. (like in the situation with nasm and this PR).

robertkirkman commented:
Regarding the viability of `-isysroot`,
I just tested these several experiments in `termux_setup_toolchain_27b.sh`

- Package used for experiments: `odt2txt`

- Experiment 1
```
export CPPFLAGS+=" -isysroot$(dirname $TERMUX_PREFIX/)"
```

```
aarch64-linux-android-clang -O2 -DHAVE_LIBZIP  -isysroot/data/data/com.termux/files  -c -o odt2txt.o odt2txt.c
aarch64-linux-android-clang -O2 -DHAVE_LIBZIP  -isysroot/data/data/com.termux/files  -c -o regex.o regex.c
aarch64-linux-android-clang -O2 -DHAVE_LIBZIP  -isysroot/data/data/com.termux/files  -c -o mem.o mem.c
aarch64-linux-android-clang -O2 -DHAVE_LIBZIP  -isysroot/data/data/com.termux/files  -c -o strbuf.o strbuf.c
odt2txt.c:20:12: fatal error: 'iconv.h' file not found
   20 | #  include <iconv.h>
      |            ^~~~~~~~~
In file included from strbuf.c:11:
./strbuf.h:15:10: fatal error: 'zlib.h' file not found
   15 | #include "zlib.h"
      |          ^~~~~~~~
In file included from regex.c:12:
In file included from ./regex.h:20:
./strbuf.h:15:10: fatal error: 'zlib.h' file not found
   15 | #include "zlib.h"
      |          ^~~~~~~~
1 error generated.
1 error generated.
make: *** [<builtin>: strbuf.o] Error 1
make: *** Waiting for unfinished jobs....
make: *** [<builtin>: regex.o] Error 1
1 error generated.
make: *** [<builtin>: odt2txt.o] Error 1
```

- Experiment 2 (same error as above)
```
export CPPFLAGS+=" -isysroot$TERMUX_PREFIX"
```

- Experiment 3 (same error as above)
```
export CPPFLAGS+=" -isysroot$TERMUX_PREFIX/include"
```

- My currently planned change (success)
```
export CPPFLAGS+=" -isystem$TERMUX_PREFIX/include"
```

Based on this result, it seems, at least to me, like `-isysroot` might not be working in this situation as well as `-isystem` does, or that maybe `-isysroot` must be used differently or used in a different place in the code, before it can work. I will keep it as `-isystem` for now and maybe someone will have a good idea in the future about how to replace it with `-isysroot` if there is a way.

robertkirkman commented:
If anyone worries that this one is taking too long, do not worry I will finish it, I just stopped uploading it because of the CI concern. 

There is a medium sized block of code in the `luarocks` package that was implementing a localized variant of the same symlink workaround I feel could be written in a globalized way, so I am combining it into my version in a way that both reduces the total number of lines of code dedicated to lua-specific symlink tasks from 12 lines to 2 lines, and prevents all of  `luarocks`' possible manifestations of the `Exec format error`. Also if I don't combine it into my version it would conflict.

robertkirkman commented:
I do not like to store too many changes only locally without backing them up in the cloud and other backups periodically, because I am afraid of storage corruption, so I am posting a snapshot of my current local changes [here](https://github.com/termux/termux-packages/compare/master...robertkirkman:termux-packages:fix-crossbuild-prefix-pollution-2), and I will probably copy and paste the changes there that are not already separated into different PRs, into other PRs, once they are ready.

I noticed from previous discussion that the granularity and documentation of any potential changes to the file `build-all.sh` is a very high priority (as opposed to making too many changes to it at once in a single PR). therefore, do not worry that the WIP `build-all.sh` I use for testing has too many lines changed in it simultaneously. I will be sure to open a separate, individual, consecutively numbered PR for every cluster of 2-3 lines changed in `build-all.sh` after it is finished, and will reimplement my own alternative implementations of the `buildorder.txt` and `buildstatus.txt` files on an as-needed basis.

twaik commented:
Probably it will be better to split this into a few small PRs. This PR already has conflicts with main branch.

robertkirkman commented:
Yes I will, but, after I ran `build-all.sh` 2x in a row (meaning, attempting to build all packages in a single builder 1 time and then after that, starting over from the first package and attempting to build all packages in the same docker container a 2nd time without deleting the docker container), I was able to successfully build most packages, but I had to skip about 300 packages due to unresolved prefix pollution, even after the current version of my changes (visible in the `fix-crossbuild-prefix-pollution-2` branch)

So, I started prioritizing the packages that can't be built even 1 single time, in a clean docker container, like `pypy3`, first, since I thought those are the higher-priority ones. A side effect of running this long 2x `build-all.sh` session using the changes from `fix-crossbuild-prefix-pollution-2`, was that at the end of it, I also obtained a list of those packages that, at the time, did not build even 1 single time in a clean docker container, and posted it in the issue.

Since then, some of those have now been fixed, so pretty soon I will check all of those packages from the shorter list of those that couldn't build at all, and see if any are still remaining that can't be built, or were moved to `disabled-packages` (like `crypto-monitor` was).

robertkirkman commented:
I have been keeping track, and everything that was involved with this PR or the linked branch is now one or more of the following:

- Merged in one of many separate PRs
- OR, has been fixed by an upstream update
- OR, the package was eventually disabled
- OR, accumulated fixes from other PRs and upstream updates eventually combined together in a cascading effect to fix packages that were not otherwise touched
- OR, should be fixed in the future on a case-by-case basis in a different way that is less messy than the original proposals

This has been superseded by newer, ongoing PRs, such as:

- https://github.com/termux/termux-packages/pull/28076
